extensions/amp-a4a/0.1/amp-a4a.js

```diff
@@ -744,6 +792,47 @@ export class AmpA4A extends AMP.BaseElement {
   }
 
   /**
+   * Refreshes ad slot by fetching a new creative and rendering it. This leaves
+   * the current creative displayed until the next one is ready.
+   *
+   * @param {function()} refreshEndCallback When called, this function will
+   *   restart the refresh cycle.
+   */
+  refresh(refreshEndCallback) {
+    this.isRefreshing = true;
+    this.tearDownSlot();
+    this.initiateAdRequest();
+    const promiseId = this.promiseId_;
+    this.adPromise_.then(() => {
+      if (!this.isRefreshing || promiseId != this.promiseId_) {
+        // If this refresh cycle was canceled, such as in a no-content
+        // response case, keep showing the old creative.
+        refreshEndCallback();
+        return;
+      }
+      this.togglePlaceholder(true);
```

Comment:
These should be within layoutCallback to ensure you are not mutating the DOM outside of a vsync.   This includes everything within the timer delay (just make sure layoutCallback returns the results of the timerFor promise.  This way you shouldn't need refresh ready promise as its all contained within layoutCallback.